platform/consul/monitor.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/golang/glog"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+type consulServices map[string][]string
+type consulServiceInstances []*api.CatalogService
+
+// Handler specifies a function to apply on an object for a given event type
+type Handler func(obj interface{}, event model.Event) error
+
+// Monitor handles service and instance changes
+type Monitor interface {
+	Start(<-chan struct{})
+	Stop()
+	AppendServiceHandler(Handler)
+	AppendInstanceHandler(Handler)
+}
+
+type consulMonitor struct {
+	discovery            *api.Client
+	instanceCachedRecord consulServiceInstances
+	serviceCachedRecord  consulServices
+	instanceHandlers     []Handler
+	serviceHandlers      []Handler
+	period               time.Duration
+	tickChan             <-chan time.Time
+	isStopped            bool
+}
+
+// NewConsulMonitor polls for changes in Consul Services and CatalogServices
+func NewConsulMonitor(client *api.Client, period time.Duration) Monitor {
+	return &consulMonitor{
+		discovery:            client,
+		period:               period,
+		instanceCachedRecord: make(consulServiceInstances, 0),
+		serviceCachedRecord:  make(consulServices),
+		instanceHandlers:     make([]Handler, 0),
+		serviceHandlers:      make([]Handler, 0),
+		isStopped:            true,
+	}
+}
+
+func (m *consulMonitor) Start(stop <-chan struct{}) {
+	m.isStopped = false
+	m.tickChan = time.NewTicker(m.period).C
+	m.run(stop)
+}
+
+func (m *consulMonitor) Stop() {
+	m.isStopped = true
+}
+
+func (m *consulMonitor) run(stop <-chan struct{}) {
+	var err error
+	for {
+		select {
+		case <-stop:
+			m.Stop()
+		case <-m.tickChan:
+			if err = m.UpdateServiceRecord(); err != nil {
+				m.Stop()
+			}
+			if err = m.UpdateInstanceRecord(); err != nil {
+				m.Stop()
+			}
+		}
+	}
+}
+
+func (m *consulMonitor) UpdateServiceRecord() error {
+	svcs, _, err := m.discovery.Catalog().Services(nil)
+	if err != nil {
+		glog.Warningf("Error:%s in fetching service result", err)
+		return err
+	}
+	newRecord := consulServices(svcs)
```

Comment:
I suggest normalizing the list by sorting it. That's a more reliable comparison.